src/libcore/ops/range.rs

```diff
@@ -326,15 +327,36 @@ impl<Idx: PartialOrd<Idx>> RangeTo<Idx> {
 /// assert_eq!(arr[1..=2], [  1,2  ]);  // RangeInclusive
 /// ```
 #[doc(alias = "..=")]
-#[derive(Clone, PartialEq, Eq, Hash)]  // not Copy -- see #27186
+#[derive(Clone)]  // not Copy -- see #27186
 #[stable(feature = "inclusive_range", since = "1.26.0")]
 pub struct RangeInclusive<Idx> {
-    // FIXME: The current representation follows RFC 1980,
-    // but it is known that LLVM is not able to optimize loops following that RFC.
-    // Consider adding an extra `bool` field to indicate emptiness of the range.
-    // See #45222 for performance test cases.
     pub(crate) start: Idx,
     pub(crate) end: Idx,
+    pub(crate) is_iterating: Option<bool>,
+    // This field is:
+    //  - `None` when next() or next_back() was never called
+    //  - `Some(true)` when `start <= end` assuming no overflow
+    //  - `Some(false)` otherwise
+    // The field cannot be a simple `bool` because the `..=` constructor can
+    // accept non-PartialOrd types, also we want the constructor to be const.
+}
+
+#[stable(feature = "inclusive_range", since = "1.26.0")]
+impl<Idx: PartialEq> PartialEq for RangeInclusive<Idx> {
+    fn eq(&self, other: &Self) -> bool {
+        self.start == other.start && self.end == other.end
```

Comment:
Are you sure you don't need to compare `is_iterating`?  Consider:

```rust
let a = 0..=0;
let mut b = a.clone();
assert_eq!(a, b);
assert_eq!(b.next(), Some(0));
assert_eq!(b.next(), None);
assert_eq!(a, b); // yes or no?
```

Currently that last assert would fail, because an exhausted range is replaced by `1..=0`.  I think with your new code, they will look equal, only different by their `is_iterating`.

Another one to consider:

```rust
let a = 1..=10;
let mut b = 0..=10;
b.next();
assert_eq!(a, b); // same `start` and `end`, different `is_iterating`
```

I guess `is_empty()` is the better value to use for `PartialEq` and `Hash`, but that needs `PartialOrd`.

Reply:
On this other hand, do we expect this to always be true?

```rust
fn reflexive(r: RangeInclusive) -> bool {
    r == (r.start()..=r.end())
}

Reply:
Imo, it would be quite jarring if it were not always true..

Reply:
Agreed with @SimonSapin and @Centril that taking `is_iterating` into account when comparing is strange. At the very least we need to modify the `impl Debug` to expose `is_iterating` if we do want to make the field significant.

Reply:
> On this other hand, do we expect this to always be true?
> 
> fn reflexive(r: RangeInclusive) -> bool {
>     r == (r.start()..=r.end())
> }

That works in the status quo because the range is changed to `1..=0` when the iterator is exhausted.  But it troubles me that with this change, `r` and `r.start()..=r.end()` might compare equal yet **not** produce the same items, if `r` is exhausted.

Reply:
Right. Both "compares equal when reconstructed form accessors" and "yields the same items if compares equal" seem desirable, but we can’t have both with this PR’s three-fields design.